### ISSUE TITLE
What should "foobar".asNumber answer?

### ISSUE THREAD

**Issue by apblack** [gracelang/language#9]
Currently, requesting `asNumber` on a string that does not represent a number answers `NaN` in the minigrace JS compiler and `0` in the minigrace C compiler.

0 is not satisfactory, because it's the right answer when the string starts with a `0`!   I think that `NaN` is actually a fine answer.  It has been suggested that an exception be raised, but in the common case where the client doesn't know whether the string represents a legal number, exceptions would become a normal flow-of-control construct, rather than an exceptional one.   Exceptions are also likely to be one or two orders of magnitude slower than normal returns.

**Comment by KimBruce** [comment]
If we leave it as NaN, can we at least add an operation that will allow us to check a value to see whether it is NaN?  The only way I know how to do it now is to request asString and then compare with "NaN", which seems annoying.

**Comment by kjx** [comment]
> On 4/08/2015, at 7:07am, Andrew Black <notifications@github.com> wrote:
> 
>  It has been suggested that an exception be raised, but in the common case where the client doesn't know whether the string represents a legal number, exceptions would become a normal flow-of-control construct, rather than an exceptional one. 

Not if there is also a String.parseAsNumberIfFail {block}   method 
String.asNumberIFail doesn’t quite feel right.

J

**Comment by apblack** [comment]
>Not if there is also a String.parseAsNumberIfFail {block}   method
>String.asNumberIFail doesn’t quite feel right.

I quite like having it return NaN.   If you want a control-flow method, how about 
```aString.ifNumber { n -> ... } else { s ->  }```
where `n` is the number, and `s` the receiver string.

**Comment by kjx** [comment]
> On 5/08/2015, at 6:40am, Andrew Black <notifications@github.com> wrote:
> 
> I quite like having it return NaN.

I hate it.
for a start: are we comitted to IEEE FP as *the* number represeentation for all of Grace.

as another data point, Kernan now uses rationals, and e.g. 1/0 throws an exception

> If you want a control-flow method, how about 
> aString.ifNumber { n -> ... } else { s -> }
> where n is the number, and s the receiver string.

sure.  actually I can’t see why the “else” brach an arugment (or at least not a mandatory one)


J

**Comment by apblack** [comment]
It's convenient for the else branch to have an argument because of
```
expensiveComputationReturningAString.ifNumber { 
    n -> ... 
} else {
    s -> SomeError.raise "The string {s} can't be converted to a number"
}

**Comment by KimBruce** [comment]
I'd replace "else" by "ifNot", but that's a minor point.  

I would have thought Andrew would be adamantly opposed to returning NaN, given his arguments about substring, etc.

If it were my choice I'd likely put in ifNumber()ifNot() as well as an asNumber that threw an exception if it wasn't a number.  This is partially because of concern about NaN always being a possible answer.

**Comment by kjx** [comment]
> On 12/01/2016, at 12:45pm, Kim Bruce <notifications@github.com> wrote:
> 
> If it were my choice I'd likely put in ifNumber()ifNot() as well as an asNumber that threw an exception if it wasn't a number. This is partially because of concern about NaN always being a possible answer.

and do we want to bake in NaNs to *all* kinds of numbers forever? 
I think propagating NaNs are bad.

**Comment by kjx** [comment]
its also convenient if blocks being passed actual arguments can ignore them  if they aren’t used, if I don’t e.g. care about the string
(or the element inside a for()do())

**Comment by KimBruce** [comment]
That I would worry about, as it might indicate a mistake that might be hard to track down.  At worst, I'd like a warning, but prefer an error.  In the case of a for-do, if you don't need the index then you probably should just use a repeat() times() to make your intention clearer.

**Comment by kjx** [comment]
Spec says Nan

**Comment by apblack** [comment]
> On 11 Jan 2016, at 21:44 , kjx <notifications@github.com> wrote:
> 
> its also convenient if blocks being passed actual arguments can ignore them if they aren’t used, if I don’t e.g. care about the string
> (or the element inside a for()do()) 

This was how minigrace used to be (and how JavaScript is by definition): excess arguments are ignored.

We agreed to change this about a year and a half ago because of negative experiences with missed errors.   I made the change, and we got some more bugs out of minigrace.

There is no good reason that I can see to change back.  Moreover, since apply, apply(_) and apply(_,_) are distinct method names, we can’t go back.

Requesting a block with the wrong number of arguments, like requesting any other method with the wrong number of arguments, is an error.   I think it’s now a “NoSuchMethod” error, but the message could, with a little work, suggest that there is a similarly named method that differs only in the number of _, between the parens.
